### PR TITLE
Transaction chain

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: src/main/scala/org/constellation/domain/consensus/ConsensusService.scala [hunk 1]
```diff
@@ -3,19 +3,20 @@ package org.constellation.domain.consensus
...
 import org.constellation.storage.algebra.{Lookup, MerkleStorageAlgebra}
 import org.constellation.storage.{ConcurrentStorageService, PendingMemPool, StorageService}
 
-abstract class ConsensusService[F[_]: Concurrent: Logger, A <: ConsensusObject]
-    extends MerkleStorageAlgebra[F, String, A] {
+abstract class ConsensusService[F[_]: Concurrent, A <: ConsensusObject] extends MerkleStorageAlgebra[F, String, A] {
 
   def metricRecordPrefix: Option[String]
 
+  private val logger = Slf4jLogger.getLogger[F]
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/consensus/ConsensusService.scala]: can't we just mix-in `with StrictLogging`?

```scala
/**
 * Defines `logger` as a value initialized with an underlying `org.slf4j.Logger`
 * named according to the class into which this trait is mixed.
 */
trait StrictLogging {

  protected val logger: Logger =
    Logger(LoggerFactory.getLogger(getClass.getName))
}
```

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala [hunk 2]
```diff
@@ -1,32 +1,63 @@
...
+      _ =>
+        ref.getUnsafe
+          .flatMap(
+            txs =>
+              (if (txs.isEmpty) {
+                 none[List[TransactionCacheData]].pure[F]
+               } else {
+                 sortForPull(txs.values.toList).flatMap { sorted =>
+                   val (left, right) = sorted.splitAt(maxCount)
+                   ref.unsafeModify(_ => (right.map(tx => tx.hash -> tx).toMap, if (left.isEmpty) none else left.some))
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala]: Alternatively to `if (left.isEmpty) none else left.some)`

```scala
Option.unless(left.isEmpty)(left)
```

https://www.scala-lang.org/api/2.13.x/scala/Option$.html#unless[A](cond:Boolean)(a:=%3EA):Option[A]

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala [hunk 3]
```diff
@@ -1,32 +1,63 @@
...
+                 }
+               })
+                .map(res => res)
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala]: redundant identity mapping? 🤔

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala [hunk 4]
```diff
@@ -1,32 +1,63 @@
...
-      .values
-      .toSeq
-      .sortBy(_.map(-_.transaction.fee.getOrElse(0L)).sum)
       .toList
-      .flatten
+      .pure[F]
+      .flatMap { t =>
+        {
+          t.traverse(
+            a =>
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala]: Can we pattern match here? I really hate `a._1` and `a._2` :D

File: src/main/scala/org/constellation/domain/transaction/TransactionService.scala [hunk 5]
```diff
@@ -19,20 +19,23 @@ import org.constellation.primitives.Schema.{
 }
 import org.constellation.primitives.{Edge, Schema, Transaction, TransactionCacheData}
 
-class TransactionService[F[_]: Concurrent: Logger](transactionChainService: TransactionChainService[F], dao: DAO)
+class TransactionService[F[_]: Concurrent](transactionChainService: TransactionChainService[F], dao: DAO)
     extends ConsensusService[F, TransactionCacheData] {
 
-  override def metricRecordPrefix: Option[String] = "Transaction".some
+  private val logger = Slf4jLogger.getLogger[F]
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/transaction/TransactionService.scala]: Same here, you can mix-in the `StrictLogging`

File: src/main/scala/org/constellation/domain/consensus/ConsensusService.scala (hunk 1 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/consensus/ConsensusService.scala]: We're using a different `Logger` than from `StrictLogging`. The logger from cats is instantiated under `F` type class and `logger.debug` is like a `Sync[F].delay { logger.debug .. }` from StrictLogging

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala (hunk 3 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala]: :+1:

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala (hunk 4 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala]: I'll try

File: src/e2e/scala/org/constellation/cluster/E2ETest.scala [hunk 6]
```diff
@@ -36,7 +36,7 @@ class E2ETest extends E2E {
   private val apis: Seq[APIClient] = nodes.map(_.getAPIClient())
   private val initialAPIs = apis
   private val addPeerRequests = nodes.map(_.getAddPeerRequest)
-  private val storeData = false
+  private val storeData = true
```
Comment (mchrapek) [inline src/e2e/scala/org/constellation/cluster/E2ETest.scala]: Should be `false`

Comment (marcinwadon) [reply src/e2e/scala/org/constellation/cluster/E2ETest.scala]: :+1:

**Pull request merged**